src/compiler/program.ts

```diff
@@ -1507,6 +1517,97 @@ namespace ts {
+
+                let foundIt = false;
+
+                // Check primary library paths
+                for (const primaryPath of primarySearchPaths) {
+                    const searchPath = combinePaths(primaryPath, ref.fileName);
+                    const resolvedFile = findLibraryDefinition(searchPath);
+                    if (resolvedFile) {
+                        processSourceFile(resolvedFile, /*isDefaultLib*/ false, /*isReference*/ true, file, ref.pos, ref.end);
+                        resolvedLibraries[ref.fileName] = { primary: true, resolvedFileName: resolvedFile };
+                        foundIt = true;
+                        break;
+                    }
+                }
+
+                // Check secondary library paths
+                if (!foundIt) {
+                    const secondaryResult = loadModuleFromNodeModules(ref.fileName, file.fileName, failedSearchPaths, moduleResolutionState);
+                    if (secondaryResult) {
+                        foundIt = true;
+                        // If we already resolved to this file, it must have been a secondary reference. Check file contents
+                        // for sameness and possibly issue an error
+                        if (previousResolution) {
+                            const otherFileText = host.readFile(secondaryResult);
+                            if (otherFileText !== getSourceFile(previousResolution.resolvedFileName).text) {
+                                fileProcessingDiagnostics.add(createFileDiagnostic(file, ref.pos, ref.end - ref.pos,
+                                    Diagnostics.Conflicting_library_definitions_for_0_found_at_1_and_2_Copy_the_correct_file_to_a_local_typings_folder_to_resolve_this_conflict,
+                                    ref.fileName,
+                                    secondaryResult,
+                                    previousResolution.resolvedFileName));
+                            }
+                        }
+                        else {
+                            // First resolution of this library
+                            processSourceFile(secondaryResult, /*isDefaultLib*/ false, /*isReference*/ true, file, ref.pos, ref.end);
+                            resolvedLibraries[ref.fileName] = { primary: false, resolvedFileName: secondaryResult };
+                        }
+                    }
+                }
+
+                if (!foundIt) {
+                    fileProcessingDiagnostics.add(createFileDiagnostic(file, ref.pos, ref.end - ref.pos, Diagnostics.Cannot_find_name_0, ref.fileName));
+                    // Create an entry as a primary lookup result so we don't keep doing this
+                    resolvedLibraries[ref.fileName] = { primary: true, resolvedFileName: undefined };
+                }
+            }
+        }
+
+        function getEffectiveLibraryPrimarySearchPaths() {
+            return options.librarySearchPaths ||
+                (options.configFilePath ?

```

Comment:
Interesting... so if the user provides `librarySearchPaths`, they effectively override all 3 of the defaults?

Also, why is the `configFilePath` added to the front? Does this mean if I have my tsconfig file in my `./src` project root, an it tries to load a `jquery` typing, it will first look for `./src/jquery/` (and see if there is a `package.json` with a `typing` field)?